File: ModernizacionPersonas.Web/ClientApp/src/app/models/grupo-asegurado.ts
import { Amparo } from '.';
import { AmparoGrupo } from './grupo-asegurado-server';

export class GrupoAsegurado {
  aseguradosOpcion1?: number;
  aseguradosOpcion2?: number;
  aseguradosOpcion3?: number;
  codigoCotizacion: number;
  codigoGrupoAsegurado?: number;
  nombreGrupoAsegurado: string;
  codigoTipoSuma?: number;
  conDistribucionAsegurados?: boolean;
  tipoSumaAsegurada?: TipoSumaAsegurada;
  valorMaxAsegurado?: number;
  valorMinAsegurado?: number;
  numeroSalariosAsegurado?: number;
  conListaAsegurados?: boolean;
  tipoEstructura?: string;
  valorAsegurado: number;
  numeroAsegurados?: number;
  edadPromedioAsegurados?: number;
  porcentajeAsegurados?: number;

  configured?: boolean;
}

export class GrupoAseguradoExtended extends GrupoAsegurado {
  codigoCotizacion: number;
  amparos: AmparoGrupo[];
  rangosGrupo: IRango[];
}

export class TipoSumaAsegurada {
  codigoTipoSumaAsegurada: number;
  nombreTipoSumaAsegurada: string;
  valorSalarioMinimo?: number;
  disabled?: boolean;
}

export interface IRango {
  codigoGrupoAsegurado?: number;
  codigoRangoGrupoAsegurado: number;
  edadMinAsegurado: number;
  edadMaxAsegurado: number;
  numeroAsegurados: number;
  valorAsegurado: number;
  isEdit?: boolean;
}

export interface Opcione {
  indiceOpcion: number;
  valorAseguradoTotal: number;
  valorAsegurado: number;
  amparo: Amparo;
  tasaRiesgo: number;
  tasaComercialAnual: number;
  porcentajeDescuento: number;
  porcentajeRecargo: number;
  tasaComercialAplicar: number;
  primaAnualIndividual: number;
  primaAnualTotal: number;
  tieneTasaSiniestralidad?: boolean;
  siniestralidad?: Siniestralidad;
  configurado: boolean;
}

export class Siniestralidad {
  amparo: Amparo;
  tasaRiesgo: number;
  tasaComercial: number;
  porcentajeDescuento: number;
  porcentajeRecargo: number;
  tasaComercialAplicar: number;
  primaAnualIndividual: number;
  primaAnualTotal: number;
}

export interface EdadIngresoPermanenciaAmparo {
  amparo: Amparo;
  edadMinimaIngreso: EdadIngresoPermanenciaAmparoField;
  edadMaximaIngreso: EdadIngresoPermanenciaAmparoField;
  edadMaximaPermanencia: EdadIngresoPermanenciaAmparoField;
  numeroDiasCarencia: EdadIngresoPermanenciaAmparoField;
}

export interface EdadIngresoPermanenciaAmparoField {
  disabled: boolean;
  rawValue?: number;
  valor: number;
}

export interface ValorAseguradoAmparo {
  amparo: Amparo;
  opciones: OpcionValorAsegurado[];
  calculoBasePorValorAsegurado: boolean;
}

export interface OpcionValorAsegurado {
  porcentajeOption?: ValorOption;
  valorOption?: ValorOption;
  primaOption?: ValorOption;
  numeroSalariosOption?: ValorOption;
  tasa?: number;
  valorDiarioOption?: ValorOption;
  numeroDiasOption?: ValorOption;
  valorDiarioDiasOption?: ValorOption;
}

export class ValorOption {
  disabled: boolean;
  rawValue?: number;
  value: string;

  static createNew(): ValorOption {
    const result: ValorOption = {
      disabled: true,
      rawValue: 0,
      value: null
    };

    return result;
  }
}

export class Asegurados {
  valorAsegurado: number;
  numeroAsegurados: number;
  aseguradosOpcion1: number;
  aseguradosOpcion2: number;
  aseguradosOpcion3: number;
  edadPromedio: number;
  porcentajeAsegurados: number;
  numeroPotencialAsegurados: number;
  conListaAsegurados: boolean;
  conDistribucionAsegurados: boolean;
  tipoEstructura: string;
  rangos: IRango[];
}

export class GrupoAseguradoCompleted {
  grupoAsegurado: GrupoAsegurado;
  selectedRamos: Amparo[];
  valoresAsegurados: ValorAseguradoAmparo[];
  edadesAmparos: EdadIngresoPermanenciaAmparo[];
  asegurados: Asegurados;
}

export enum ValoresAseguradosEvents {
  BASE_CALCULO_CHANGE = 'BASE_CALCULO_CHANGE',
  PRIMA_VALUE_CHANGE = 'PRIMA_VALUE_CHANGE',
  PERCENT_VALUE_CHANGE = 'PERCENT_VALUE_CHANGE',
  SALARIO_VALUE_CHANGE = 'SALARIO_VALUE_CHANGE',
  CURRENCY_VALUE_CHANGE = 'CURRENCY_VALUE_CHANGE',
  DIAS_VALUE_CHANGE = 'DIAS_VALUE_CHANGE',
  VALOR_DIARIO_CHANGE= 'VALOR_DIARIO_CHANGE'
}

export interface OpcionesUsadas {
  opcion1: boolean;
  opcion2: boolean;
  opcion3: boolean;
}

File: ModernizacionPersonas.Entities/GrupoAsegurado.cs
using System.Collections.Generic;

namespace ModernizacionPersonas.Entities
{
    public class GrupoAsegurado
    {
        public int CodigoCotizacion { get; set; }
        public int CodigoGrupoAsegurado { get; set; }
        public string NombreGrupoAsegurado { get; set; }
        public int CodigoTipoSuma { get; set; }
        public decimal ValorMinAsegurado { get; set; }
        public decimal ValorMaxAsegurado { get; set; }
        public int NumeroSalariosAsegurado { get; set; }
        public List<AmparoGrupoAsegurado> AmparosGrupo { get; set; }
        //public bool ConListaAsegurados { get; set; }
        public bool ConDistribucionAsegurados { get; set; }
        public decimal ValorAsegurado { get; set; }
        public int NumeroAsegurados { get; set; }
        public int AseguradosOpcion1 { get; set; }
        public int AseguradosOpcion2 { get; set; }
        public int AseguradosOpcion3 { get; set; }
        public int EdadPromedioAsegurados { get; set; }
        public decimal PorcentajeAsegurados { get; set; }
        public int NumeroPotencialAsegurados { get; set; }
        public int TipoEstructura { get; set; }
        public IEnumerable<Rango> RangosGrupo { get; set; }
        public GrupoAsegurado()
        {
            this.AmparosGrupo = new List<AmparoGrupoAsegurado>();
            this.RangosGrupo = new List<Rango>();
        }
    }

    class GrupoAseguradoViewModel
    {
        public int CodigoCotizacion { get; set; }
        public int CodigoGrupoAsegurado { get; set; }
        public string NombreGrupoAsegurado { get; set; }
        public int CodigoTipoSuma { get; set; }
        // public TipoSumaAsegurada TipoSumaAsegurada { get; set; }
        public decimal ValorMinAsegurado { get; set; }
        public decimal ValorMaxAsegurado { get; set; }
        public int NumeroSalariosAsegurado { get; set; }
        public bool ConListaAsegurados { get; set; }
        public decimal ValorAsegurado { get; set; }
        public int NumeroAsegurados { get; set; }
        public decimal PorcentajeAsegurados { get; set; }
    }
}
